## This is a notebook for the Toronto's neighborhoods for coursera capstone course

In [187]:
import numpy as np
import pandas as pd
import matplotlib as plt

In [188]:
df=pd.read_csv(r'C:\Users\dee\Documents\Python\Data Science Labs\Capstone Project\Toronto_Neighborhoud.csv')

In [189]:
df.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

In [190]:
df.shape

(180, 3)

### In the next cell we are dropping the Borough with value 'Not assigned'

In [191]:
df.drop(df.index[df['Borough']=='Not assigned'],inplace=True)

In [192]:
df.reset_index(drop=True,inplace=True)

In [193]:
pd.set_option('display.max_rows', 105)
df.head(100)

Postal Code           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   
12         M1C       Scarborough   
13         M3C        North York   
14         M4C         East York   
15         M5C  Downtown Toronto   
16         M6C              York   
17         M9C         Etobicoke   
18         M1E       Scarborough   
19         M4E      East Toronto   
20         M5E  Downtown Toronto   
21         M6E              York   
22         M1G       Scarborough   
23         M4G         East York   
24         M5G  Downtown Toronto   
25         M6G  Downtown Toronto   
26         M1H       Scarborough   
27         M2H        North York   
28         M3H        North York   
29         M4H         East York   
30         M5H  Downtown Toronto   
31         M6H      West Toronto   
32         M1J       Scarborough   
33         M2J        North York   
34         M3J        North York   
35         M4J         East York   
36         M5J  Downtown Toronto   
37         M6J      West Toronto   
38         M1K       Scarborough   
39         M2K        North York   
40         M3K        North York   
41         M4K      East Toronto   
42         M5K  Downtown Toronto   
43         M6K      West Toronto   
44         M1L       Scarborough   
45         M2L        North York   
46         M3L        North York   
47         M4L      East Toronto   
48         M5L  Downtown Toronto   
49         M6L        North York   
50         M9L        North York   
51         M1M       Scarborough   
52         M2M        North York   
53         M3M        North York   
54         M4M      East Toronto   
55         M5M        North York   
56         M6M              York   
57         M9M        North York   
58         M1N       Scarborough   
59         M2N        North York   
60         M3N        North York   
61         M4N   Central Toronto   
62         M5N   Central Toronto   
63         M6N              York   
64         M9N              York   
65         M1P       Scarborough   
66         M2P        North York   
67         M4P   Central Toronto   
68         M5P   Central Toronto   
69         M6P      West Toronto   
70         M9P         Etobicoke   
71         M1R       Scarborough   
72         M2R        North York   
73         M4R   Central Toronto   
74         M5R   Central Toronto   
75         M6R      West Toronto   
76         M7R       Mississauga   
77         M9R         Etobicoke   
78         M1S       Scarborough   
79         M4S   Central Toronto   
80         M5S  Downtown Toronto   
81         M6S      West Toronto   
82         M1T       Scarborough   
83         M4T   Central Toronto   
84         M5T  Downtown Toronto   
85         M1V       Scarborough   
86         M4V   Central Toronto   
87         M5V  Downtown Toronto   
88         M8V         Etobicoke   
89         M9V         Etobicoke   
90         M1W       Scarborough   
91         M4W  Downtown Toronto   
92         M5W  Downtown Toronto   
93         M8W         Etobicoke   
94         M9W         Etobicoke   
95         M1X       Scarborough   
96         M4X  Downtown Toronto   
97         M5X  Downtown Toronto   
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   

                                        Neighbourhood  
0                                           Parkwoods  
1                                    Victoria Village  
2                           Regent Park, Harbourfront  
3                    Lawrence Manor, Lawrence Heights  
4         Queen's Park, Ontario Provincial Government  
5             Islington Avenue

In [194]:
df.shape

(103, 3)

In [195]:
df.groupby(['Borough']).count()

Postal Code  Neighbourhood
Borough                                     
Central Toronto             9              9
Downtown Toronto           19             19
East Toronto                5              5
East York                   5              5
Etobicoke                  12             12
Mississauga                 1              1
North York                 24             24
Scarborough                17             17
West Toronto                6              6
York                        5              5

In [196]:
df['Borough'].value_counts()

North York          24
Downtown Toronto    19
Scarborough         17
Etobicoke           12
Central Toronto      9
West Toronto         6
East York            5
East Toronto         5
York                 5
Mississauga          1
Name: Borough, dtype: int64

### Check if there is a Neighbourhood which is 'Not assigned'

In [197]:
df[df['Neighbourhood']=='Not assigned'].count()

Postal Code      0
Borough          0
Neighbourhood    0
dtype: int64

### Last we have to check if there more than one neighbourhoods for the same Postal Code

In [198]:
df.groupby(['Postal Code']).count()

Borough  Neighbourhood
Postal Code                        
M1B                1              1
M1C                1              1
M1E                1              1
M1G                1              1
M1H                1              1
M1J                1              1
M1K                1              1
M1L                1              1
M1M                1              1
M1N                1              1
M1P                1              1
M1R                1              1
M1S                1              1
M1T                1              1
M1V                1              1
M1W                1              1
M1X                1              1
M2H                1              1
M2J                1              1
M2K                1              1
M2L                1              1
M2M                1              1
M2N                1              1
M2P                1              1
M2R                1              1
M3A                1              1
M3B                1              1
M3C                1              1
M3H                1              1
M3J                1              1
M3K                1              1
M3L                1              1
M3M                1              1
M3N                1              1
M4A                1              1
M4B                1              1
M4C                1              1
M4E                1              1
M4G                1              1
M4H                1              1
M4J                1              1
M4K                1              1
M4L                1              1
M4M                1              1
M4N                1              1
M4P                1              1
M4R                1              1
M4S                1              1
M4T                1              1
M4V                1              1
M4W                1              1
M4X                1              1
M4Y                1              1
M5A                1              1
M5B                1              1
M5C                1              1
M5E                1              1
M5G                1              1
M5H                1              1
M5J                1              1
M5K                1              1
M5L                1              1
M5M                1              1
M5N                1              1
M5P                1              1
M5R                1              1
M5S                1              1
M5T                1              1
M5V                1              1
M5W                1              1
M5X                1              1
M6A                1              1
M6B                1              1
M6C                1              1
M6E                1              1
M6G                1              1
M6H                1              1
M6J                1              1
M6K                1              1
M6L                1              1
M6M                1              1
M6N                1              1
M6P                1              1
M6R                1              1
M6S                1              1
M7A                1              1
M7R                1              1
M7Y                1              1
M8V                1              1
M8W                1              1
M8X                1              1
M8Y                1              1
M8Z                1              1
M9A                1              1
M9B                1              1
M9C                1              1
M9L                1              1
M9M                1              1
M9N                1              1
M9P                1              1
M9R                1              1
M9V                1              1
M9W                1              1

In [199]:
df.shape

(103, 3)

In [200]:
aa=df['Postal Code'].unique()
len(aa)

103

### We see that count() and len of unique values of 'Postal Code' have 103 rows  same as the shape
### As a result each Postal Code is listed once

### We can also see that the example given in the instructions is already fixed in Wikipedia

In [201]:
df[df['Postal Code']=='M5A']

Postal Code           Borough              Neighbourhood
2         M5A  Downtown Toronto  Regent Park, Harbourfront

In [202]:
df.head()

Postal Code           Borough                                Neighbourhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [203]:
df.shape

(103, 3)

In [204]:
dg=pd.read_csv(r'C:\Users\dee\Documents\Python\Data Science Labs\Capstone Project\Geospatial_Coordinates.csv')

In [205]:
dg.head(100)

Postal Code   Latitude  Longitude
0          M1B  43.806686 -79.194353
1          M1C  43.784535 -79.160497
2          M1E  43.763573 -79.188711
3          M1G  43.770992 -79.216917
4          M1H  43.773136 -79.239476
5          M1J  43.744734 -79.239476
6          M1K  43.727929 -79.262029
7          M1L  43.711112 -79.284577
8          M1M  43.716316 -79.239476
9          M1N  43.692657 -79.264848
10         M1P  43.757410 -79.273304
11         M1R  43.750072 -79.295849
12         M1S  43.794200 -79.262029
13         M1T  43.781638 -79.304302
14         M1V  43.815252 -79.284577
15         M1W  43.799525 -79.318389
16         M1X  43.836125 -79.205636
17         M2H  43.803762 -79.363452
18         M2J  43.778517 -79.346556
19         M2K  43.786947 -79.385975
20         M2L  43.757490 -79.374714
21         M2M  43.789053 -79.408493
22         M2N  43.770120 -79.408493
23         M2P  43.752758 -79.400049
24         M2R  43.782736 -79.442259
25         M3A  43.753259 -79.329656
26         M3B  43.745906 -79.352188
27         M3C  43.725900 -79.340923
28         M3H  43.754328 -79.442259
29         M3J  43.767980 -79.487262
30         M3K  43.737473 -79.464763
31         M3L  43.739015 -79.506944
32         M3M  43.728496 -79.495697
33         M3N  43.761631 -79.520999
34         M4A  43.725882 -79.315572
35         M4B  43.706397 -79.309937
36         M4C  43.695344 -79.318389
37         M4E  43.676357 -79.293031
38         M4G  43.709060 -79.363452
39         M4H  43.705369 -79.349372
40         M4J  43.685347 -79.338106
41         M4K  43.679557 -79.352188
42         M4L  43.668999 -79.315572
43         M4M  43.659526 -79.340923
44         M4N  43.728020 -79.388790
45         M4P  43.712751 -79.390197
46         M4R  43.715383 -79.405678
47         M4S  43.704324 -79.388790
48         M4T  43.689574 -79.383160
49         M4V  43.686412 -79.400049
50         M4W  43.679563 -79.377529
51         M4X  43.667967 -79.367675
52         M4Y  43.665860 -79.383160
53         M5A  43.654260 -79.360636
54         M5B  43.657162 -79.378937
55         M5C  43.651494 -79.375418
56         M5E  43.644771 -79.373306
57         M5G  43.657952 -79.387383
58         M5H  43.650571 -79.384568
59         M5J  43.640816 -79.381752
60         M5K  43.647177 -79.381576
61         M5L  43.648198 -79.379817
62         M5M  43.733283 -79.419750
63         M5N  43.711695 -79.416936
64         M5P  43.696948 -79.411307
65         M5R  43.672710 -79.405678
66         M5S  43.662696 -79.400049
67         M5T  43.653206 -79.400049
68         M5V  43.628947 -79.394420
69         M5W  43.646435 -79.374846
70         M5X  43.648429 -79.382280
71         M6A  43.718518 -79.464763
72         M6B  43.709577 -79.445073
73         M6C  43.693781 -79.428191
74         M6E  43.689026 -79.453512
75         M6G  43.669542 -79.422564
76         M6H  43.669005 -79.442259
77         M6J  43.647927 -79.419750
78         M6K  43.636847 -79.428191
79         M6L  43.713756 -79.490074
80         M6M  43.691116 -79.476013
81         M6N  43.673185 -79.487262
82         M6P  43.661608 -79.464763
83         M6R  43.648960 -79.456325
84         M6S  43.651571 -79.484450
85         M7A  43.662301 -79.389494
86         M7R  43.636966 -79.615819
87         M7Y  43.662744 -79.321558
88         M8V  43.605647 -79.501321
89         M8W  43.602414 -79.543484
90         M8X  43.653654 -79.506944
91         M8Y  43.636258 -79.498509
92         M8Z  43.628841 -79.520999
93         M9A  43.667856 -79.532242
94         M9B  43.650943 -79.554724
95         M9C  43.643515 -79.577201
96         M9L  43.756303 -79.565963
97         M9M  43.724766 -79.532242
98         M9N  43.706876 -79.518188
99         M9P  43.696319 -79.532242

In [206]:
df=df.sort_values('Postal Code')

In [207]:
df.sort_values('Postal Code',inplace=True)

In [208]:
df.reset_index(drop=True,inplace=True)

In [209]:
df.head(100)

Postal Code           Borough  \
0          M1B       Scarborough   
1          M1C       Scarborough   
2          M1E       Scarborough   
3          M1G       Scarborough   
4          M1H       Scarborough   
5          M1J       Scarborough   
6          M1K       Scarborough   
7          M1L       Scarborough   
8          M1M       Scarborough   
9          M1N       Scarborough   
10         M1P       Scarborough   
11         M1R       Scarborough   
12         M1S       Scarborough   
13         M1T       Scarborough   
14         M1V       Scarborough   
15         M1W       Scarborough   
16         M1X       Scarborough   
17         M2H        North York   
18         M2J        North York   
19         M2K        North York   
20         M2L        North York   
21         M2M        North York   
22         M2N        North York   
23         M2P        North York   
24         M2R        North York   
25         M3A        North York   
26         M3B        North York   
27         M3C        North York   
28         M3H        North York   
29         M3J        North York   
30         M3K        North York   
31         M3L        North York   
32         M3M        North York   
33         M3N        North York   
34         M4A        North York   
35         M4B         East York   
36         M4C         East York   
37         M4E      East Toronto   
38         M4G         East York   
39         M4H         East York   
40         M4J         East York   
41         M4K      East Toronto   
42         M4L      East Toronto   
43         M4M      East Toronto   
44         M4N   Central Toronto   
45         M4P   Central Toronto   
46         M4R   Central Toronto   
47         M4S   Central Toronto   
48         M4T   Central Toronto   
49         M4V   Central Toronto   
50         M4W  Downtown Toronto   
51         M4X  Downtown Toronto   
52         M4Y  Downtown Toronto   
53         M5A  Downtown Toronto   
54         M5B  Downtown Toronto   
55         M5C  Downtown Toronto   
56         M5E  Downtown Toronto   
57         M5G  Downtown Toronto   
58         M5H  Downtown Toronto   
59         M5J  Downtown Toronto   
60         M5K  Downtown Toronto   
61         M5L  Downtown Toronto   
62         M5M        North York   
63         M5N   Central Toronto   
64         M5P   Central Toronto   
65         M5R   Central Toronto   
66         M5S  Downtown Toronto   
67         M5T  Downtown Toronto   
68         M5V  Downtown Toronto   
69         M5W  Downtown Toronto   
70         M5X  Downtown Toronto   
71         M6A        North York   
72         M6B        North York   
73         M6C              York   
74         M6E              York   
75         M6G  Downtown Toronto   
76         M6H      West Toronto   
77         M6J      West Toronto   
78         M6K      West Toronto   
79         M6L        North York   
80         M6M              York   
81         M6N              York   
82         M6P      West Toronto   
83         M6R      West Toronto   
84         M6S      West Toronto   
85         M7A  Downtown Toronto   
86         M7R       Mississauga   
87         M7Y      East Toronto   
88         M8V         Etobicoke   
89         M8W         Etobicoke   
90         M8X         Etobicoke   
91         M8Y         Etobicoke   
92         M8Z         Etobicoke   
93         M9A         Etobicoke   
94         M9B         Etobicoke   
95         M9C         Etobicoke   
96         M9L        North York   
97         M9M        North York   
98         M9N              York   
99         M9P         Etobicoke   

                                        Neighbourhood  
0                                      Malvern, Rouge  
1              Rouge Hill, Port Union, Highland Creek  
2                   Guildwood, Morningside, West Hill  
3                                              Woburn  
4                                           Cedarbrae  
5                             

In [210]:
df['Latitude']=dg['Latitude']

In [211]:
df['Longitude']=dg['Longitude']

In [212]:
df.head(104)

Postal Code           Borough  \
0           M1B       Scarborough   
1           M1C       Scarborough   
2           M1E       Scarborough   
3           M1G       Scarborough   
4           M1H       Scarborough   
5           M1J       Scarborough   
6           M1K       Scarborough   
7           M1L       Scarborough   
8           M1M       Scarborough   
9           M1N       Scarborough   
10          M1P       Scarborough   
11          M1R       Scarborough   
12          M1S       Scarborough   
13          M1T       Scarborough   
14          M1V       Scarborough   
15          M1W       Scarborough   
16          M1X       Scarborough   
17          M2H        North York   
18          M2J        North York   
19          M2K        North York   
20          M2L        North York   
21          M2M        North York   
22          M2N        North York   
23          M2P        North York   
24          M2R        North York   
25          M3A        North York   
26          M3B        North York   
27          M3C        North York   
28          M3H        North York   
29          M3J        North York   
30          M3K        North York   
31          M3L        North York   
32          M3M        North York   
33          M3N        North York   
34          M4A        North York   
35          M4B         East York   
36          M4C         East York   
37          M4E      East Toronto   
38          M4G         East York   
39          M4H         East York   
40          M4J         East York   
41          M4K      East Toronto   
42          M4L      East Toronto   
43          M4M      East Toronto   
44          M4N   Central Toronto   
45          M4P   Central Toronto   
46          M4R   Central Toronto   
47          M4S   Central Toronto   
48          M4T   Central Toronto   
49          M4V   Central Toronto   
50          M4W  Downtown Toronto   
51          M4X  Downtown Toronto   
52          M4Y  Downtown Toronto   
53          M5A  Downtown Toronto   
54          M5B  Downtown Toronto   
55          M5C  Downtown Toronto   
56          M5E  Downtown Toronto   
57          M5G  Downtown Toronto   
58          M5H  Downtown Toronto   
59          M5J  Downtown Toronto   
60          M5K  Downtown Toronto   
61          M5L  Downtown Toronto   
62          M5M        North York   
63          M5N   Central Toronto   
64          M5P   Central Toronto   
65          M5R   Central Toronto   
66          M5S  Downtown Toronto   
67          M5T  Downtown Toronto   
68          M5V  Downtown Toronto   
69          M5W  Downtown Toronto   
70          M5X  Downtown Toronto   
71          M6A        North York   
72          M6B        North York   
73          M6C              York   
74          M6E              York   
75          M6G  Downtown Toronto   
76          M6H      West Toronto   
77          M6J      West Toronto   
78          M6K      West Toronto   
79          M6L        North York   
80          M6M              York   
81          M6N              York   
82          M6P      West Toronto   
83          M6R      West Toronto   
84          M6S      West Toronto   
85          M7A  Downtown Toronto   
86          M7R       Mississauga   
87          M7Y      East Toronto   
88          M8V         Etobicoke   
89          M8W         Etobicoke   
90          M8X         Etobicoke   
91          M8Y         Etobicoke   
92          M8Z         Etobicoke   
93          M9A         Etobicoke   
94          M9B         Etobicoke   
95          M9C         Etobicoke   
96          M9L        North York   
97          M9M        North York   
98          M9N              York   
99          M9P         Etobicoke   
100         M9R         Etobicoke   
101         M9V         Etobicoke   
102         M9W         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                       Malvern, Rouge  43.806686 -79.194353

In [213]:
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

In [214]:
address = 'Toronto, Canada'
#address = 'Langstrasse'
#address = 'Paralimni'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of New York City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of New York City are 43.6534817, -79.3839347.


In [215]:
import folium

In [216]:
#create Toronto map
map_Toronto=folium.Map(location=[latitude,longitude],zoom_start=10.5)
#add markers
for lat,long,borough,neighbourhood in zip (df['Latitude'],df['Longitude'],df['Borough'],df['Neighbourhood']):
    label = '{borough},{neighbourhood}'
    label = folium.Popup(label,parse_html=True)
    folium.CircleMarker(
    [lat,long],
    radius=5,
    color='red',
    fill=True,
    fill_color='#FF4500',
    fill_opacity=0.7,
    parse_html=False).add_to(map_Toronto)

map_Toronto

### Chose to analyze the neighbourhoods in Central Toronto, Downtown Toronto York and North York

In [217]:
df_new=df[(df['Borough']=='York') | (df['Borough']=='Central Toronto')| (df['Borough']=='Downtown Toronto') | (df['Borough']=='North York')]
df_new.reset_index(drop=True,inplace=True)
df_new

Postal Code           Borough  \
0          M2H        North York   
1          M2J        North York   
2          M2K        North York   
3          M2L        North York   
4          M2M        North York   
5          M2N        North York   
6          M2P        North York   
7          M2R        North York   
8          M3A        North York   
9          M3B        North York   
10         M3C        North York   
11         M3H        North York   
12         M3J        North York   
13         M3K        North York   
14         M3L        North York   
15         M3M        North York   
16         M3N        North York   
17         M4A        North York   
18         M4N   Central Toronto   
19         M4P   Central Toronto   
20         M4R   Central Toronto   
21         M4S   Central Toronto   
22         M4T   Central Toronto   
23         M4V   Central Toronto   
24         M4W  Downtown Toronto   
25         M4X  Downtown Toronto   
26         M4Y  Downtown Toronto   
27         M5A  Downtown Toronto   
28         M5B  Downtown Toronto   
29         M5C  Downtown Toronto   
30         M5E  Downtown Toronto   
31         M5G  Downtown Toronto   
32         M5H  Downtown Toronto   
33         M5J  Downtown Toronto   
34         M5K  Downtown Toronto   
35         M5L  Downtown Toronto   
36         M5M        North York   
37         M5N   Central Toronto   
38         M5P   Central Toronto   
39         M5R   Central Toronto   
40         M5S  Downtown Toronto   
41         M5T  Downtown Toronto   
42         M5V  Downtown Toronto   
43         M5W  Downtown Toronto   
44         M5X  Downtown Toronto   
45         M6A        North York   
46         M6B        North York   
47         M6C              York   
48         M6E              York   
49         M6G  Downtown Toronto   
50         M6L        North York   
51         M6M              York   
52         M6N              York   
53         M7A  Downtown Toronto   
54         M9L        North York   
55         M9M        North York   
56         M9N              York   

                                        Neighbourhood   Latitude  Longitude  
0                                   Hillcrest Village  43.803762 -79.363452  
1                        Fairview, Henry Farm, Oriole  43.778517 -79.346556  
2                                     Bayview Village  43.786947 -79.385975  
3                            York Mills, Silver Hills  43.757490 -79.374714  
4                             Willowdale, Newtonbrook  43.789053 -79.408493  
5                         Willowdale, Willowdale East  43.770120 -79.408493  
6                                     York Mills West  43.752758 -79.400049  
7                         Willowdale, Willowdale West  43.782736 -79.442259  
8                                           Parkwoods  43.753259 -79.329656  
9                                           Don Mills  43.745906 -79.352188  
10                                          Don Mills  43.725900 -79.340923  
11    Bathurst Manor, Wilson Heights, Downsview North  43.754328 -79.442259  
12                    Northwood Park, York University  43.767980 -79.487262  
13                                          Downsview  43.737473 -79.464763  
14                                          Downsview  43.739015 -79.506944  
15                                          Downsview  43.728496 -79.495697  
16                                          Downsview  43.761631 -79.520999  
17                                   Victoria Village  43.725882 -79.315572  
18                                      Lawrence Park  43.728020 -79.388790  
19                                   Davisville North  43.712751 -79.390197  
20                  North Toronto West, Lawrence Park  43.715383 -79.405678  
21                                         Davisville  43.704324 -79.388790  
22                        Moore Park, Summerhill East  43.689574 -79.383160  
23  Summerhill West, Rathnelly, South Hill

In [218]:
#create Toronto map
map_NorthYork=folium.Map(location=[latitude,longitude],zoom_start=10.5)
#add markers
for lat,long,borough,neighbourhood in zip (df_new['Latitude'],df_new['Longitude'],df_new['Borough'],df_new['Neighbourhood']):
    label = '{borough},{neighbourhood}'
    label = folium.Popup(label,parse_html=True)
    folium.CircleMarker(
    [lat,long],
    radius=5,
    color='yellow',
    fill=True,
    fill_color='#FF4500',
    fill_opacity=0.7,
    parse_html=False).add_to(map_NorthYork)

map_NorthYork

In [219]:
CLIENT_ID = 'S0FKAQ0NPT4Z2U1N1YPVM5G44JPXEEE0YW0UAGG1V4OLEQAG' # your Foursquare ID
CLIENT_SECRET = '0JGQISOOFFFGWCGXKPHKEAMCO0V3YBHIQUG0OUIQJDTHUQKN' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: S0FKAQ0NPT4Z2U1N1YPVM5G44JPXEEE0YW0UAGG1V4OLEQAG
CLIENT_SECRET:0JGQISOOFFFGWCGXKPHKEAMCO0V3YBHIQUG0OUIQJDTHUQKN


In [220]:
df_new.describe()

Latitude  Longitude
count  57.000000  57.000000
mean   43.705730 -79.413216
std     0.046356   0.054082
min    43.628947 -79.565963
25%    43.662696 -79.442259
50%    43.706876 -79.400049
75%    43.739015 -79.379817
max    43.803762 -79.315572

In [221]:
used_lat=df_new['Latitude'].mean()
used_long=df_new['Longitude'].mean()
print(f"Latitude: \t {used_lat}\nLongitude: \t{used_long}")

Latitude: 	 43.705729942105265
Longitude: 	-79.41321637017545


### Let's get the some venues around this Location in a radius of 500 meters

In [222]:
# type your answer here
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius
# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    used_lat, 
    used_long, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=S0FKAQ0NPT4Z2U1N1YPVM5G44JPXEEE0YW0UAGG1V4OLEQAG&client_secret=0JGQISOOFFFGWCGXKPHKEAMCO0V3YBHIQUG0OUIQJDTHUQKN&v=20180605&ll=43.705729942105265,-79.41321637017545&radius=500&limit=100'

In [223]:
import requests

In [224]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5fa7cdb8589bc022c5abf1c1'},
 'response': {'headerLocation': 'Toronto',
  'headerFullLocation': 'Toronto',
  'headerLocationGranularity': 'city',
  'totalResults': 21,
  'suggestedBounds': {'ne': {'lat': 43.71022994660527,
    'lng': -79.40700304247484},
   'sw': {'lat': 43.70122993760526, 'lng': -79.41942969787605}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4fdc0e98e4b05197cd14912b',
       'name': 'The Abbot',
       'location': {'address': '508 Eglinton Ave W',
        'crossStreet': 'Heddington Ave',
        'lat': 43.703687730512435,
        'lng': -79.41348481516249,
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.703687730512435,
          'lng': -79.41348481516249}],
        'distance': 228,
        '

## Now lets explore the neighbourhoods in Central Toronto, Downtoan Toronto and York

In [225]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [226]:
Toronto_venues = getNearbyVenues(names=df_new['Neighbourhood'],
                                   latitudes=df_new['Latitude'],
                                   longitudes=df_new['Longitude'])

Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
York Mills, Silver Hills
Willowdale, Newtonbrook
Willowdale, Willowdale East
York Mills West
Willowdale, Willowdale West
Parkwoods
Don Mills
Don Mills
Bathurst Manor, Wilson Heights, Downsview North
Northwood Park, York University
Downsview
Downsview
Downsview
Downsview
Victoria Village
Lawrence Park
Davisville North
North Toronto West, Lawrence Park
Davisville
Moore Park, Summerhill East
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
Rosedale
St. James Town, Cabbagetown
Church and Wellesley
Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
Bedford Park, Lawrence Manor East
Roselawn
Forest Hill North & West, Forest Hill Road Park
The Annex, North Midtown, Yorkville
University of Toronto, Harbord
Kensington Market, 

In [227]:
Toronto_venues.shape

(1612, 7)

In [228]:
Toronto_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Hillcrest Village              43.803762              -79.363452   
1  Hillcrest Village              43.803762              -79.363452   
2  Hillcrest Village              43.803762              -79.363452   
3  Hillcrest Village              43.803762              -79.363452   
4  Hillcrest Village              43.803762              -79.363452   

                                 Venue  Venue Latitude  Venue Longitude  \
0               Eagle's Nest Golf Club       43.805455       -79.364186   
1                      AY Jackson Pool       43.804515       -79.366138   
2                         Villa Madina       43.801685       -79.363938   
3                    Duncan Creek Park       43.805539       -79.360695   
4  A.Y. Jackson Secondary School Track       43.805068       -79.366677   

             Venue Category  
0               Golf Course  
1                      Pool  
2  Mediterranean Restaurant  
3                   Dog Run  
4        Athletics & Sports

In [229]:
Toronto_venues.groupby(by='Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Bathurst Manor, Wilson Heights, Downsview North                        21   
Bayview Village                                                         4   
Bedford Park, Lawrence Manor East                                      22   
Berczy Park                                                            55   
CN Tower, King and Spadina, Railway Lands, Harb...                     16   
Caledonia-Fairbanks                                                     4   
Central Bay Street                                                     68   
Christie                                                               16   
Church and Wellesley                                                   75   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             33   
Davisville North                                                        9   
Del Ray, Mount Dennis, Keelsdale and Silverthorn                        5   
Don Mills                                                              23   
Downsview                                                              16   
Fairview, Henry Farm, Oriole                                           70   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         4   
Garden District, Ryerson                                              100   
Glencairn                                                               5   
Harbourfront East, Union Station, Toronto Islands                     100   
Hillcrest Village                                                       5   
Humber Summit                                                           2   
Humberlea, Emery                                                        1   
Humewood-Cedarvale                                                      5   
Kensington Market, Chinatown, Grange Park                              74   
Lawrence Manor, Lawrence Heights                                       13   
Lawrence Park                                                           3   
Moore Park, Summerhill East                                             2   
North Park, Maple Leaf Park, Upwood Park                                3   
North Toronto West, Lawrence Park                                      18   
Northwood Park, York University                                         7   
Parkwoods                                                               2   
Queen's Park, Ontario Provincial Government                            33   
Regent Park, Harbourfront                                              44   
Richmond, Adelaide, King                                              100   
Rosedale                                                                4   
Roselawn                                                                2   
Runnymede, The Junction North                                           4   
St. James Town                                                         85   
St. James Town, Cabbagetown                                            48   
Stn A PO Boxes                                                         96   
Summerhill West, Rathnelly, South Hill, Forest ...                     14   
The Annex, North Midtown, Yorkville                                    19   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         34   
Victoria Village                                                        5   
Weston                                                                  2   
Willowdale, Willowdale East                                            33   
Willowdale, Willowdale West                                             5   
York Mills West                                 

In [230]:
Toronto_venues['Neighborhood'].value_counts()

Toronto Dominion Centre, Design Exchange                                                                      100
Richmond, Adelaide, King                                                                                      100
Commerce Court, Victoria Hotel                                                                                100
First Canadian Place, Underground city                                                                        100
Harbourfront East, Union Station, Toronto Islands                                                             100
Garden District, Ryerson                                                                                      100
Stn A PO Boxes                                                                                                 96
St. James Town                                                                                                 85
Church and Wellesley                                                                    

### Now let's get dummies for the categories of the venue to better analyze the neighbourhoods

In [231]:
# one hot encoding
Toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
Toronto_onehot['Neighborhood'] = Toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
cols = list(Toronto_onehot.columns.values) #Make a list of all of the columns in the df
cols.pop(cols.index('Neighborhood')) #Remove columns from list
Toronto_onehot = Toronto_onehot[['Neighborhood']+cols] #Create new dataframe with columns in the order you want
Toronto_onehot.head()

pd.set_option('display.max_columns', 225)
Toronto_onehot.head()

Neighborhood  Accessories Store  Afghan Restaurant  Airport  \
0  Hillcrest Village                  0                  0        0   
1  Hillcrest Village                  0                  0        0   
2  Hillcrest Village                  0                  0        0   
3  Hillcrest Village                  0                  0        0   
4  Hillcrest Village                  0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0           0                    0                 0                   0   
1           0                    0                 0                   0   
2           0                    0                 0                   0   
3           0                    0                 0                   0   
4           0                    0                 0                   1   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0          0           0           0       0     0    0               0   
1          0           0           0       0     0    0               0   
2          0           0           0       0     0    0               0   
3          0           0           0       0     0    0               0   
4          0           0           0       0     0    0               0   

   Baseball Stadium  Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  \
0                 0                   0      0                0         0   
1                 0                   0      0                0         0   
2                 0                   0      0                0         0   
3                 0                   0      0                0         0   
4                 0                   0      0                0         0   

   Beer Store  Belgian Restaurant  Bike Shop  Bistro  Boat or Ferry  \
0           0                   0          0       0              0   
1           0                   0          0       0              0   
2           0                   0          0       0              0   
3           0                   0          0       0              0   
4           0                   0          0       0              0   

   Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  \
0          0         0                     0               0        0   
1          0         0                     0               0        0   
2          0         0                     0               0        0   
3          0         0                     0               0        0   
4          0         0                     0               0        0   

   Bridal Shop  Bubble Tea Shop  Building  Burger Joint  Burrito Place  \
0            0                0         0             0              0   
1            0                0         0             0              0   
2            0                0         0             0              0   
3            0                0         0             0              0   
4            0                0         0             

In [232]:
Toronto_onehot.shape

(1612, 239)

In [233]:
Toronto_grouped= Toronto_onehot.groupby('Neighborhood').mean().reset_index()
Toronto_grouped

Neighborhood  Accessories Store  \
0     Bathurst Manor, Wilson Heights, Downsview North           0.000000   
1                                     Bayview Village           0.000000   
2                   Bedford Park, Lawrence Manor East           0.000000   
3                                         Berczy Park           0.000000   
4   CN Tower, King and Spadina, Railway Lands, Har...           0.000000   
5                                 Caledonia-Fairbanks           0.000000   
6                                  Central Bay Street           0.000000   
7                                            Christie           0.000000   
8                                Church and Wellesley           0.000000   
9                      Commerce Court, Victoria Hotel           0.000000   
10                                         Davisville           0.000000   
11                                   Davisville North           0.000000   
12   Del Ray, Mount Dennis, Keelsdale and Silverthorn           0.000000   
13                                          Don Mills           0.000000   
14                                          Downsview           0.000000   
15                       Fairview, Henry Farm, Oriole           0.000000   
16             First Canadian Place, Underground city           0.000000   
17    Forest Hill North & West, Forest Hill Road Park           0.000000   
18                           Garden District, Ryerson           0.000000   
19                                          Glencairn           0.000000   
20  Harbourfront East, Union Station, Toronto Islands           0.000000   
21                                  Hillcrest Village           0.000000   
22                                      Humber Summit           0.000000   
23                                   Humberlea, Emery           0.000000   
24                                 Humewood-Cedarvale           0.000000   
25          Kensington Market, Chinatown, Grange Park           0.000000   
26                   Lawrence Manor, Lawrence Heights           0.076923   
27                                      Lawrence Park           0.000000   
28                        Moore Park, Summerhill East           0.000000   
29           North Park, Maple Leaf Park, Upwood Park           0.000000   
30                  North Toronto West, Lawrence Park           0.000000   
31                    Northwood Park, York University           0.000000   
32                                          Parkwoods           0.000000   
33        Queen's Park, Ontario Provincial Government           0.000000   
34                          Regent Park, Harbourfront           0.000000   
35                           Richmond, Adelaide, King           0.000000   
36                                           Rosedale           0.000000   
37                                           Roselawn           0.000000   
38                      Runnymede, The Junction North           0.000000   
39                                     St. James Town           0.000000   
40                        St. James Town, Cabbagetown           0.000000   
41                                     Stn A PO Boxes           0.000000   
42  Summerhill West, Rathnelly, South Hill, Forest...           0.000000   
43                The Annex, North Midtown, Yorkville           0.000000   
44           Toronto Dominion Centre, Design Exchange           0.000000   
45                     University of Toronto, Harbord           0.000000   
46                                   Victoria Village           0.000000   
47                                             Weston           0.000000   
48                        Willowdale, Willowdale East           0.000000   
49                        Willowdale, Willowdale West           0.000000   
50                                    York Mills West           0.000000   
51                           York Mills, Silver Hills           0.000000   

    Afghan R

### Next we are taking the top 10 venues for each neighbourhood

In [234]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [235]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = Toronto_grouped['Neighborhood']

for ind in np.arange(Toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0    Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
1                                    Bayview Village    Chinese Restaurant   
2                  Bedford Park, Lawrence Manor East           Coffee Shop   
3                                        Berczy Park           Coffee Shop   
4  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Bank         Shopping Mall              Pharmacy   
1   Japanese Restaurant                  Bank                  Café   
2        Sandwich Place    Italian Restaurant             Juice Bar   
3           Cheese Shop                Bakery          Cocktail Bar   
4       Airport Service         Boat or Ferry       Harbor / Marina   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                 Diner            Restaurant           Bridal Shop   
1           Yoga Studio          Dessert Shop     Electronics Store   
2     Indian Restaurant                   Pub               Butcher   
3    Seafood Restaurant              Beer Bar            Restaurant   
4   Rental Car Location                 Plane           Coffee Shop   

         8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                Deli / Bodega           Supermarket         Ice Cream Shop  
1  Eastern European Restaurant   Dumpling Restaurant             Donut Shop  
2                         Café      Sushi Restaurant            Pizza Place  
3               Farmers Market        Breakfast Spot                  Beach  
4                          Bar              Boutique       Airport Terminal

In [236]:
neighborhoods_venues_sorted.shape

(52, 11)

### Finaly is time to cluster our neighbourhoods

In [237]:
from sklearn.cluster import KMeans

In [238]:
# set number of clusters
kclusters = 5

Toronto_grouped_clustering = Toronto_grouped.drop('Neighborhood', 1) #axis nomizw to 1
Toronto_grouped_clustering
# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0)
kmeans.fit(Toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:100] 

array([1, 1, 1, 1, 1, 4, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1,
       1, 3, 1, 1, 1, 4, 0, 4, 1, 1, 4, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 4, 1, 1, 4, 2])

### And visualize the clusters

In [239]:
Toronto_grouped

Neighborhood  Accessories Store  \
0     Bathurst Manor, Wilson Heights, Downsview North           0.000000   
1                                     Bayview Village           0.000000   
2                   Bedford Park, Lawrence Manor East           0.000000   
3                                         Berczy Park           0.000000   
4   CN Tower, King and Spadina, Railway Lands, Har...           0.000000   
5                                 Caledonia-Fairbanks           0.000000   
6                                  Central Bay Street           0.000000   
7                                            Christie           0.000000   
8                                Church and Wellesley           0.000000   
9                      Commerce Court, Victoria Hotel           0.000000   
10                                         Davisville           0.000000   
11                                   Davisville North           0.000000   
12   Del Ray, Mount Dennis, Keelsdale and Silverthorn           0.000000   
13                                          Don Mills           0.000000   
14                                          Downsview           0.000000   
15                       Fairview, Henry Farm, Oriole           0.000000   
16             First Canadian Place, Underground city           0.000000   
17    Forest Hill North & West, Forest Hill Road Park           0.000000   
18                           Garden District, Ryerson           0.000000   
19                                          Glencairn           0.000000   
20  Harbourfront East, Union Station, Toronto Islands           0.000000   
21                                  Hillcrest Village           0.000000   
22                                      Humber Summit           0.000000   
23                                   Humberlea, Emery           0.000000   
24                                 Humewood-Cedarvale           0.000000   
25          Kensington Market, Chinatown, Grange Park           0.000000   
26                   Lawrence Manor, Lawrence Heights           0.076923   
27                                      Lawrence Park           0.000000   
28                        Moore Park, Summerhill East           0.000000   
29           North Park, Maple Leaf Park, Upwood Park           0.000000   
30                  North Toronto West, Lawrence Park           0.000000   
31                    Northwood Park, York University           0.000000   
32                                          Parkwoods           0.000000   
33        Queen's Park, Ontario Provincial Government           0.000000   
34                          Regent Park, Harbourfront           0.000000   
35                           Richmond, Adelaide, King           0.000000   
36                                           Rosedale           0.000000   
37                                           Roselawn           0.000000   
38                      Runnymede, The Junction North           0.000000   
39                                     St. James Town           0.000000   
40                        St. James Town, Cabbagetown           0.000000   
41                                     Stn A PO Boxes           0.000000   
42  Summerhill West, Rathnelly, South Hill, Forest...           0.000000   
43                The Annex, North Midtown, Yorkville           0.000000   
44           Toronto Dominion Centre, Design Exchange           0.000000   
45                     University of Toronto, Harbord           0.000000   
46                                   Victoria Village           0.000000   
47                                             Weston           0.000000   
48                        Willowdale, Willowdale East           0.000000   
49                        Willowdale, Willowdale West           0.000000   
50                                    York Mills West           0.000000   
51                           York Mills, Silver Hills           0.000000   

    Afghan R

In [240]:
df_new.shape

(57, 5)

In [242]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
neighborhoods_venues_sorted
Toronto_merged = df_new

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
Toronto_merged = Toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighbourhood')

Toronto_merged.head(52) # check the last columns!

Postal Code           Borough  \
0          M2H        North York   
1          M2J        North York   
2          M2K        North York   
3          M2L        North York   
4          M2M        North York   
5          M2N        North York   
6          M2P        North York   
7          M2R        North York   
8          M3A        North York   
9          M3B        North York   
10         M3C        North York   
11         M3H        North York   
12         M3J        North York   
13         M3K        North York   
14         M3L        North York   
15         M3M        North York   
16         M3N        North York   
17         M4A        North York   
18         M4N   Central Toronto   
19         M4P   Central Toronto   
20         M4R   Central Toronto   
21         M4S   Central Toronto   
22         M4T   Central Toronto   
23         M4V   Central Toronto   
24         M4W  Downtown Toronto   
25         M4X  Downtown Toronto   
26         M4Y  Downtown Toronto   
27         M5A  Downtown Toronto   
28         M5B  Downtown Toronto   
29         M5C  Downtown Toronto   
30         M5E  Downtown Toronto   
31         M5G  Downtown Toronto   
32         M5H  Downtown Toronto   
33         M5J  Downtown Toronto   
34         M5K  Downtown Toronto   
35         M5L  Downtown Toronto   
36         M5M        North York   
37         M5N   Central Toronto   
38         M5P   Central Toronto   
39         M5R   Central Toronto   
40         M5S  Downtown Toronto   
41         M5T  Downtown Toronto   
42         M5V  Downtown Toronto   
43         M5W  Downtown Toronto   
44         M5X  Downtown Toronto   
45         M6A        North York   
46         M6B        North York   
47         M6C              York   
48         M6E              York   
49         M6G  Downtown Toronto   
50         M6L        North York   
51         M6M              York   

                                        Neighbourhood   Latitude  Longitude  \
0                                   Hillcrest Village  43.803762 -79.363452   
1                        Fairview, Henry Farm, Oriole  43.778517 -79.346556   
2                                     Bayview Village  43.786947 -79.385975   
3                            York Mills, Silver Hills  43.757490 -79.374714   
4                             Willowdale, Newtonbrook  43.789053 -79.408493   
5                         Willowdale, Willowdale East  43.770120 -79.408493   
6                                     York Mills West  43.752758 -79.400049   
7                         Willowdale, Willowdale West  43.782736 -79.442259   
8                                           Parkwoods  43.753259 -79.329656   
9                                           Don Mills  43.745906 -79.352188   
10                                          Don Mills  43.725900 -79.340923   
11    Bathurst Manor, Wilson Heights, Downsview North  43.754328 -79.442259   
12                    Northwood Park, York University  43.767980 -79.487262   
13                                          Downsview  43.737473 -79.464763   
14                                          Downsview  43.739015 -79.506944   
15                                          Downsview  43.728496 -79.495697   
16                                          Downsview  43.761631 -79.520999   
17                                   Victoria Village  43.725882 -79.315572   
18                                      Lawrence Park  43.728020 -79.388790   
19                                   Davisville North  43.712751 -79.390197   
20                  North Toronto West, Lawrence Park  43.715383 -79.405678   
21                                         Davisville  43.704324 -79.388790   
22                        Moore Park, Summerhill East  43.689574 -79.383160   
23  Summerhill West, Rathnelly, South Hill, Forest...  43.686412 -79.400049   
24                                           Rosedale  43.679563 -79.377529   
25                        St. James Town

In [246]:
Toronto_merged.dropna(inplace=True) #default axis=0 diladi fevgei rows

In [247]:
Toronto_merged['Cluster Labels']=Toronto_merged['Cluster Labels'].astype(int)

In [248]:
Toronto_merged.head(10)

Postal Code     Borough                 Neighbourhood   Latitude  \
0          M2H  North York             Hillcrest Village  43.803762   
1          M2J  North York  Fairview, Henry Farm, Oriole  43.778517   
2          M2K  North York               Bayview Village  43.786947   
3          M2L  North York      York Mills, Silver Hills  43.757490   
5          M2N  North York   Willowdale, Willowdale East  43.770120   
6          M2P  North York               York Mills West  43.752758   
7          M2R  North York   Willowdale, Willowdale West  43.782736   
8          M3A  North York                     Parkwoods  43.753259   
9          M3B  North York                     Don Mills  43.745906   
10         M3C  North York                     Don Mills  43.725900   

    Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0  -79.363452               1    Athletics & Sports           Golf Course   
1  -79.346556               1        Clothing Store           Coffee Shop   
2  -79.385975               1    Chinese Restaurant   Japanese Restaurant   
3  -79.374714               2   Martial Arts School           Yoga Studio   
5  -79.408493               1      Ramen Restaurant           Pizza Place   
6  -79.400049               4                  Park     Convenience Store   
7  -79.442259               1           Coffee Shop               Butcher   
8  -79.329656               4                  Park     Food & Drink Shop   
9  -79.352188               1                   Gym            Beer Store   
10 -79.340923               1                   Gym            Beer Store   

   3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
0                   Pool               Dog Run     Mediterranean Restaurant   
1   Fast Food Restaurant         Women's Store                    Juice Bar   
2                   Bank                  Café                  Yoga Studio   
3       Department Store     Electronics Store  Eastern European Restaurant   
5       Sushi Restaurant        Sandwich Place          Japanese Restaurant   
6          Deli / Bodega     Electronics Store  Eastern European Restaurant   
7               Pharmacy           Pizza Place                Grocery Store   
8   Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   
9    Japanese Restaurant           Coffee Shop                    Bike Shop   
10   Japanese Restaurant           Coffee Shop                    Bike Shop   

   6th Most Common Venue 7th Most Common Venue        8th Most Common Venue  \
0         Discount Store      Department Store                 Dessert Shop   
1    Japanese Restaurant                Bakery                   Restaurant   
2           Dessert Shop     Electronics Store  Eastern European Restaurant   
3    Dumpling Restaurant            Donut Shop             Doner Restaurant   
5                   Café           Coffee Shop                    Pet Store   
6    Dumpling Restaurant            Donut Shop             Doner Restaurant   
7           Airport Gate           Event Space                  Escape Room   
8    Dumpling Restaurant            Donut Shop             Doner Restaurant   
9   Caribbean Restaurant            Restaurant           Italian Restaurant   
10  Caribbean Restaurant            Restaurant           Italian Restaurant   

   9th Most Common Venue       10th Most Common Venue  
0     Dim Sum Restaurant                        Diner  
1             Shoe Store               Cosmetics Shop  
2    Dumpling Restaurant                   Donut Shop  
3                Dog Run          Distribution Center  
5                  Plaza                       Lounge  
6                Dog Run          Distribution Center  
7      Electronics Store  Eastern European Restaurant  
8                Dog Run          Distribution Center  
9    Sporting Goods Shop               Discount Store  
10   Sporting Goods Shop               Discount Store

## MAP

In [243]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [249]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Toronto_merged['Latitude'], Toronto_merged['Longitude'], Toronto_merged['Neighbourhood'], Toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[(cluster)-1],
        fill=True,
        fill_color=rainbow[(cluster)-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [ ]:
rainbow[0]

In [ ]:
cluster